crates/telemetry/src/traces.rs

```diff
@@ -0,0 +1,180 @@
+use std::str::FromStr;
+
+use opentelemetry::{
+    global,
+    propagation::{Extractor, Injector},
+    KeyValue,
+};
+use opentelemetry_otlp::WithExportConfig;
+use opentelemetry_otlp::{OTEL_EXPORTER_OTLP_ENDPOINT, OTEL_EXPORTER_OTLP_TRACES_ENDPOINT};
+use opentelemetry_sdk::{trace::Tracer, Resource};
+use opentelemetry_semantic_conventions::resource::SERVICE_NAME;
+use tracing::{metadata::LevelFilter, Level};
+use tracing_opentelemetry::{OpenTelemetryLayer, OpenTelemetrySpanExt};
+use tracing_subscriber::{Layer, Registry};
+
+use crate::config::Config;
+
+/// Constructs a layer for the tracing subscriber that sends spans to an OTEL collector.
+///
+/// In addition to the settings provided by [Config] it also pulls OTEL configuration from the
+/// environment based on the variables defined
+/// [here](https://opentelemetry.io/docs/specs/otel/protocol/exporter/).
+pub(crate) fn otel_tracing_layer(
+    config: Config,
+) -> anyhow::Result<
+    tracing_subscriber::filter::Filtered<
+        OpenTelemetryLayer<Registry, Tracer>,
+        LevelFilter,
+        Registry,
+    >,
+> {
+    let mut service_metadata = vec![KeyValue::new(SERVICE_NAME, config.otel_service_name)];
+    service_metadata.extend(config.otel_resource_attributes.inner());
+
+    // This will configure the exporter based on the OTEL_EXPORTER_* environment variables. We
+    // currently default to using the HTTP exporter but in the future we could select off of the
+    // combination of OTEL_EXPORTER_OTLP_PROTOCOL and OTEL_EXPORTER_OTLP_TRACES_PROTOCOL to
+    // determine whether we should use http/protobuf or grpc.
+    let mut exporter = opentelemetry_otlp::new_exporter().http();
+
+    // This mitigation was taken from https://github.com/neondatabase/neon/blob/main/libs/tracing-utils/src/lib.rs
+    //
+    // opentelemetry-otlp v0.15.0 has a bug in how it uses the
+    // OTEL_EXPORTER_OTLP_ENDPOINT env variable. According to the
+    // OpenTelemetry spec at
+    // <https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#endpoint-urls-for-otlphttp>,
+    // the full exporter URL is formed by appending "/v1/traces" to the value
+    // of OTEL_EXPORTER_OTLP_ENDPOINT. However, opentelemetry-otlp only does
+    // that with the grpc-tonic exporter. Other exporters, like the HTTP
+    // exporter, use the URL from OTEL_EXPORTER_OTLP_ENDPOINT as is, without
+    // appending "/v1/traces".
+    //
+    // See https://github.com/open-telemetry/opentelemetry-rust/pull/950
+    //
+    // Work around that by checking OTEL_EXPORTER_OTLP_ENDPOINT, and setting
+    // the endpoint url with the "/v1/traces" path ourselves. If the bug is
+    // fixed in a later version, we can remove this code. But if we don't
+    // remember to remove this, it won't do any harm either, as the crate will
+    // just ignore the OTEL_EXPORTER_OTLP_ENDPOINT setting when the endpoint
+    // is set directly with `with_endpoint`.
```

Comment:
This is good context but it breaks up the flow of reading this function. I'd recommend extracting the mitigation (and comment) into a e.g. `fix_endpoint_bug(endpoint: &mut String)` function.